// @jlguenego/asn.1/build/src/analysis/ASN1Lexer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ASN1Lexer = exports.allASN1Tokens = exports.NumberToken = exports.NegativeNumberToken = exports.Identifier = exports.TypeReference = exports.RANGE_SEPARATOR = exports.PIPE = exports.SEMI_COLUMN = exports.COMMA = exports.R_SQUARE = exports.L_SQUARE = exports.R_PARENTHESIS = exports.L_PARENTHESIS = exports.R_CURLY = exports.L_CURLY = exports.AFFECTATION = exports.WhiteSpace = exports.Comment = void 0;
const chevrotain_1 = require("chevrotain");
const ASN1Keyword_1 = require("./lexer/ASN1Keyword");
exports.Comment = (0, chevrotain_1.createToken)({
    name: 'Comment',
    pattern: /--.*?(?:\n|--)/,
    group: chevrotain_1.Lexer.SKIPPED,
});
exports.WhiteSpace = (0, chevrotain_1.createToken)({
    name: 'WhiteSpace',
    pattern: /[ \t\n\r]+/,
    group: chevrotain_1.Lexer.SKIPPED,
});
// Keywords
exports.AFFECTATION = (0, chevrotain_1.createToken)({ name: 'AFFECTATION', pattern: /::=/ });
exports.L_CURLY = (0, chevrotain_1.createToken)({ name: 'LCurly', pattern: /{/ });
exports.R_CURLY = (0, chevrotain_1.createToken)({ name: 'RCurly', pattern: /}/ });
exports.L_PARENTHESIS = (0, chevrotain_1.createToken)({ name: 'LParenthesis', pattern: /\(/ });
exports.R_PARENTHESIS = (0, chevrotain_1.createToken)({ name: 'RParenthesis', pattern: /\)/ });
exports.L_SQUARE = (0, chevrotain_1.createToken)({ name: 'LSquareBracket', pattern: /\[/ });
exports.R_SQUARE = (0, chevrotain_1.createToken)({ name: 'RSquareBracket', pattern: /\]/ });
exports.COMMA = (0, chevrotain_1.createToken)({ name: 'Comma', pattern: /,/ });
exports.SEMI_COLUMN = (0, chevrotain_1.createToken)({ name: 'SemiColumn', pattern: /;/ });
exports.PIPE = (0, chevrotain_1.createToken)({ name: 'Pipe', pattern: /\|/ });
exports.RANGE_SEPARATOR = (0, chevrotain_1.createToken)({
    name: 'RangeSeparator',
    pattern: /[.][.]/,
});
exports.TypeReference = (0, chevrotain_1.createToken)({
    name: 'TypeReference',
    // 12.2.1
    // Note: use of non capturing group. (?:xxx)
    pattern: /[A-Z](?:-?\w+)*/,
});
exports.Identifier = (0, chevrotain_1.createToken)({
    name: 'Identifier',
    // 12.3
    pattern: /[a-z](?:-?\w+)*/,
});
exports.NegativeNumberToken = (0, chevrotain_1.createToken)({
    name: 'NegativeNumber',
    // 12.8 and 19.1
    pattern: /-[1-9][0-9]*/,
});
exports.NumberToken = (0, chevrotain_1.createToken)({
    name: 'Number',
    // 12.8
    pattern: /(?:[1-9][0-9]*)|0/,
});
exports.allASN1Tokens = [
    exports.Comment,
    exports.WhiteSpace,
    ...Object.values(ASN1Keyword_1.k),
    // 3 chars
    exports.AFFECTATION,
    // 2 chars
    exports.RANGE_SEPARATOR,
    // 1 chars
    exports.L_CURLY,
    exports.R_CURLY,
    exports.L_PARENTHESIS,
    exports.R_PARENTHESIS,
    exports.L_SQUARE,
    exports.R_SQUARE,
    exports.COMMA,
    exports.SEMI_COLUMN,
    exports.PIPE,
    // The Identifier must appear after the keywords because all keywords are valid identifiers.
    exports.Identifier,
    exports.TypeReference,
    exports.NegativeNumberToken,
    exports.NumberToken,
];
exports.ASN1Lexer = new chevrotain_1.Lexer(exports.allASN1Tokens);
