// @jlguenego/asn.1/build/src/analysis/lexer/ASN1Keyword.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.k = void 0;
const chevrotain_1 = require("chevrotain");
const token = (0, chevrotain_1.createToken)({
    name: 'x',
    pattern: /x/,
});
exports.k = {
    // 'ENCODING-CONTROL': token,
    // 'RELATIVE-OID-IRI': token,
    // 'ABSTRACT-SYNTAX': token,
    // 'TYPE-IDENTIFIER': token,
    // 'MINUS-INFINITY': token,
    // 'PLUS-INFINITY': token,
    // 'NOT-A-NUMBER': token,
    // 'RELATIVE-OID': token,
    // 'TIME-OF-DAY': token,
    // 'DATE-TIME': token,
    // 'OID-IRI': token,
    ObjectDescriptor: token,
    GeneralizedTime: token,
    PrintableString: token,
    UniversalString: token,
    VideotexString: token,
    EXTENSIBILITY: token,
    GeneralString: token,
    GraphicString: token,
    NumericString: token,
    TeletexString: token,
    VisibleString: token,
    INSTRUCTIONS: token,
    INTERSECTION: token,
    ISO646String: token,
    APPLICATION: token,
    CONSTRAINED: token,
    DEFINITIONS: token,
    COMPONENTS: token,
    CONTAINING: token,
    ENUMERATED: token,
    IDENTIFIER: token,
    UTF8String: token,
    AUTOMATIC: token,
    BMPString: token,
    CHARACTER: token,
    COMPONENT: token,
    IA5String: token,
    T61String: token,
    UNIVERSAL: token,
    DURATION: token,
    EMBEDDED: token,
    EXPLICIT: token,
    EXTERNAL: token,
    IMPLICIT: token,
    INCLUDES: token,
    INSTANCE: token,
    OPTIONAL: token,
    SEQUENCE: token,
    SETTINGS: token,
    BOOLEAN: token,
    DEFAULT: token,
    ENCODED: token,
    EXPORTS: token,
    IMPLIED: token,
    IMPORTS: token,
    INTEGER: token,
    PATTERN: token,
    PRESENT: token,
    PRIVATE: token,
    UTCTime: token,
    ABSENT: token,
    EXCEPT: token,
    OBJECT: token,
    CHOICE: token,
    STRING: token,
    SYNTAX: token,
    UNIQUE: token,
    CLASS: token,
    BEGIN: token,
    FALSE: token,
    OCTET: token,
    UNION: token,
    DATE: token,
    FROM: token,
    SIZE: token,
    REAL: token,
    TAGS: token,
    TIME: token,
    TRUE: token,
    WITH: token,
    ALL: token,
    END: token,
    MAX: token,
    SET: token,
    MIN: token,
    BIT: token,
    NULL: token,
    PDV: token,
    BY: token,
    OF: token,
};
for (const name of Object.keys(exports.k)) {
    exports.k[name] = (0, chevrotain_1.createToken)({
        name,
        pattern: new RegExp(name),
    });
}
