// @jlguenego/asn.1/build/test/generate.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const fs_1 = require("fs");
const path_1 = require("path");
const src_1 = require("../src");
describe('Generate Test', () => {
    it('should generate a message', () => {
        // ASN1 module definition parser.
        const asn1ModuleStr = `
HelloProtocol DEFINITIONS ::= BEGIN

    Person ::= SEQUENCE {
        lastname           GeneralString,
        firstname          GeneralString,
        likeCoding     BOOLEAN
    }

    WelcomeMsg ::= SEQUENCE {
        id        INTEGER,
        someone   Person
    }

END
`;
        const data = {
            id: 34,
            someone: {
                lastname: 'GUENEGO',
                firstname: 'Jean-Louis',
                likeCoding: true,
            },
        };
        const asn1Module = src_1.ASN1.getModuleFromStr(asn1ModuleStr);
        const buffer = src_1.ASN1.generate(asn1Module, 'WelcomeMsg', data, {
            encodingRule: src_1.EncodingRule.DER,
        });
        const messageHex = buffer.toString('hex');
        const expectedMessageHex = (0, src_1.sanitize)((0, fs_1.readFileSync)((0, path_1.resolve)(__dirname, 'data/hello-jlg.der'), {
            encoding: 'utf8',
        }));
        assert_1.default.deepStrictEqual(messageHex, expectedMessageHex);
    });
});
