// node-expose-sspi/dist/sso/mutex.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.activeDirectoryMutex = exports.Mutex = void 0;
const debug_1 = __importDefault(require("debug"));
const debug = (0, debug_1.default)('node-expose-sspi:mutex');
class Mutex {
    constructor() {
        this.isBusy = false;
        this.queue = [];
    }
    onRelease() {
        debug('release');
        if (this.queue.length === 0) {
            this.isBusy = false;
            return;
        }
        const { resolve } = this.queue.shift();
        debug('decrease queue size', this.queue.length);
        this.isBusy = true;
        resolve(this.onRelease.bind(this));
    }
    async acquire() {
        return new Promise((resolve) => {
            debug('acquire');
            if (!this.isBusy) {
                this.isBusy = true;
                return resolve(this.onRelease.bind(this));
            }
            this.queue.push({ resolve });
            debug('increase queue size', this.queue.length);
        });
    }
}
exports.Mutex = Mutex;
exports.activeDirectoryMutex = new Mutex();
